File: sdk/python/src/freewayai/utilities.py
import json

def _format_messages(prompt, messages, variables):

    formatted_messages = []

    for message in messages:
        formatted_messages.append({
            "role": message["role"],
            "content": _format_template_string(prompt, message["content"], variables)
        })

    return formatted_messages

def _read_prompt_from_system_and_id(system_id, systems_location, prompt_id):
    with open(f'{systems_location}/{system_id}/prompt_templates.json') as f:
        prompts = json.load(f)
        target_prompt = prompts["templates"][prompt_id]
        return target_prompt
    
def _read_query_from_system_and_id(system_id, systems_location, query_id):
    with open(f'{systems_location}/{system_id}/database_templates.json') as f:
        prompts = json.load(f)
        target_prompt = prompts["templates"][query_id]
        return target_prompt

def _format_template_string(body, message_template, variables):
    needed_keys = [entry["name"] for entry in body["variables"]]
    for key in needed_keys:
        if key not in variables:
            raise ValueError(f"Missing required key {key}")
    
    return message_template.format(**variables)

File: sdk/js/freewayai/src/utilities.ts
import fs from "fs";

function _read_prompt_from_system_and_id(systemId: string, systemsLocation: string, promptId: string): any {
  const filePath = `${systemsLocation}/${systemId}/prompt_templates.json`;
  const fileContent = fs.readFileSync(filePath, 'utf8');
  const prompts = JSON.parse(fileContent);
  const targetPrompt = prompts["templates"][promptId];
  return targetPrompt;
}

function _read_query_from_system_and_id(systemId: string, systemsLocation: string, queryId: string): any {
  const filePath = `${systemsLocation}/${systemId}/database_templates.json`;
  const fileContent = fs.readFileSync(filePath, 'utf8');
  const queries = JSON.parse(fileContent);
  const targetQuery = queries["templates"][queryId];
  return targetQuery;
}

function _format_template_string(body: any, messageTemplate: string, variables: Record<string, any>): string {
  const neededKeys = body["variables"].map((entry: any) => entry["name"]);
  for (const key of neededKeys) {
    if (!(key in variables)) {
      throw new Error(`Missing required key ${key}`);
    }
  }

  return messageTemplate.replace(/\{([^}]+)\}/g, (match, key) => {
    if (key in variables) {
      return variables[key];
    } else {
      throw new Error(`Missing value for key ${key}`);
    }
  });
}

export { 
  _read_prompt_from_system_and_id, 
  _read_query_from_system_and_id, 
  _format_template_string 
};
